lock-impl/src/main/java/com/palantir/lock/impl/LockServiceImpl.java

```diff
@@ -1061,10 +1061,12 @@ private StringBuilder getGeneralLockStats() {
 
     @Override
     public void close() {
+        if (!isShutDown) {
+            executor.shutdownNow();
+            wakeIndefiniteBlockers();
+            callOnClose.run();
+        }
         isShutDown = true;
-        executor.shutdownNow();
-        wakeIndefiniteBlockers();
-        callOnClose.run();
     }
```

Comment:
Hmm. This generally fixes the issue, but is not thread-safe; if two threads call `close()` concurrently both could enter the if-block, and `callOnClose` is an arbitrary Runnable so not necessarily idempotent.

Minimal fix could be to synchronize this (people shouldn't depend on close being super performant). A better fix could actually be the AtomicBoolean idea you mention!